inspire_json_merger/utils/filterout_utils.py

```diff
@@ -59,24 +64,67 @@ def filter_conflicts_by_path(conflict_list, to_delete_path):
 
 
 def is_to_delete(conflict, keys_path):
-    to_delete = keys_path.split('.')
-    conflict_path = conflict_to_list(conflict)
+    conflict_path = '.'.join(conflict_to_list(conflict))
+    return conflict_path.startswith(keys_path)
 
-    if len(to_delete) > len(conflict_path):
-        return False
 
-    i = 0
-    while i < len(conflict_path) and i < len(to_delete):
-        c = conflict_path[i]
-        d = to_delete[i]
+def conflict_to_list(conflict):
+    path = conflict[1]
+    return [p for p in path if not isinstance(p, int)]
 
-        if c != d:
-            return False
-        i += 1
 
-    return True
+def _remove_source_from_field(source, field):
+    """Remove all elements matching ``source`` in ``field``."""
+    return [element for element in field if element.get('source') != source]
 
 
-def conflict_to_list(conflict):
-    path = conflict[1]
-    return [p for p in path if not isinstance(p, int)]
+def keep_only_update_source_in_field(field, root, head, update):
+    """Remove elements from root and head where ``source`` matches the update.
+
+    This is useful if the update needs to overwrite all elements with the same
+    source.
+
+    .. note::
+        If the update doesn't contain exactly one source in ``field``, the
+        records are returned with no modifications.
+
+    Args:
+        field(str): the field to filter out.
+        root(dict): the root record, whose ``field`` will be cleaned.
+        head(dict): the head record, whose ``field`` will be cleaned.
+        update(dict): the update record, from which the ``source`` is read.
+
+    Returns:
+        tuple: ``(root, head, update)`` with some elements filtered out from
+            ``root`` and ``head``.
+    """
+    update_sources = set(get_value(update, '.'.join([field, 'source']), []))
+    if len(update_sources) != 1:
+        return root, head, update
+    source = update_sources.pop()
+
+    root = freeze(root)
+    head = freeze(head)
+    if field in root:
+        root = root.set(field, _remove_source_from_field(source, root[field]))
+    if field in head:
+        head = head.set(field, _remove_source_from_field(source, head[field]))
+
+    return thaw(root), thaw(head), update
+
+
+def filter_records(root, head, update, filters=()):
```

Comment:
I find the empty tuple syntax a bit confusing (`('snth')` is actually a string, but `()` is a tuple...) but it's an interesting alternative to the usual:
```python
def something(maybelist=None):
    maybelist = maybelist or []
    for item in maybelist:
    ....
```

Reply:
yes, an empty tuple is iterable and immutable, so it's better to use it instead of `None` when you are iterating over the argument.

Reply:
using `()` vs `tuple()` is also 3 to 5 times faster on both python2 and 3... for some reason :/

```
dcaro@pcrcssis001$ python3
Python 3.6.2 (default, Oct  2 2017, 16:51:32) 
[GCC 7.2.1 20170915 (Red Hat 7.2.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import timeit
>>> timeit.timeit(stmt='a=tuple()') 
0.10025268299796153
>>> timeit.timeit(stmt='a=()')
0.021887153998250142


dcaro@pcrcssis001$ python
Python 2.7.13 (default, Sep  5 2017, 08:53:59) 
[GCC 7.1.1 20170622 (Red Hat 7.1.1-3)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import timeit
>>> timeit.timeit(stmt='a=tuple()') 
0.06886100769042969
>>> timeit.timeit(stmt='a=()')
0.019394874572753906
```